### ISSUE TITLE
Decorated test methods don't run when trial'ed directly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3909]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#3909|
|Type|defect|
|Created|2009-07-10 06:38:03Z|
|Branch|https://github.com/twisted/twisted/tree/trial-decorated-methods-3909|

When using decorated tests:

```python
from twisted.trial import unittest

print "Print works in this environment."

def a_decorator(fn):
        print 'Decorating', fn 
        def newfunc(*args, **kwargs):
                print "I am decorated."
                fn(*args, **kwargs)
        return newfunc

class TestA(unittest.TestCase):
        def test_something(self):
                self.assertEqual(1, 1)
        test_something = a_decorator(test_something)
```

`trial decorated_test.TestA` works, but `trial decorated_test.TestA.test_something` does not:

```
# trial decorated_test.TestA.test_something
Print works in this environment.
Decorating <function test_something at 0x7f101729d5f0>
Traceback (most recent call last):
  File "/usr/local/bin/trial", line 22, in <module>
    run()
  File "/opt/Python-latest/lib/python2.7/site-packages/twisted/scripts/trial.py", line 328, in run
    suite = _getSuite(config)
  File "/opt/Python-latest/lib/python2.7/site-packages/twisted/scripts/trial.py", line 286, in _getSuite
    return loader.loadByNames(config['tests'], recurse)
  File "/opt/Python-latest/lib/python2.7/site-packages/twisted/trial/runner.py", line 657, in loadByNames
    for thing in set(things)]
  File "/opt/Python-latest/lib/python2.7/site-packages/twisted/trial/runner.py", line 621, in loadAnything
    return self.loadMethod(thing)
  File "/opt/Python-latest/lib/python2.7/site-packages/twisted/trial/runner.py", line 524, in loadMethod
    return self._makeCase(method.im_class, method.__name__)
  File "/opt/Python-latest/lib/python2.7/site-packages/twisted/trial/runner.py", line 527, in _makeCase
    return klass(methodName)
  File "/opt/Python-latest/lib/python2.7/site-packages/twisted/trial/unittest.py", line 672, in __init__
    super(TestCase, self).__init__(methodName)
  File "/opt/Python-latest/lib/python2.7/unittest.py", line 363, in __init__
    (self.__class__, methodName))
ValueError: no such test method in <class 'decorated_test.TestA'>: newfunc

```

Attachments:

* [misnamed-method.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/84b/84b28e67dac3802a6dc50b12b3010df4a0a937b7/ebebcab2de01217e37bbd0d8088147aa4ec87b1d.patch) (5381 bytes) - added by kelly on 2009-10-17 23:35:56Z - 
* [misnamed-method-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/84b/84b28e67dac3802a6dc50b12b3010df4a0a937b7/1b4bfc5fc82b380417b751f70d7d6567b38839f3.patch) (10040 bytes) - added by kelly on 2009-10-23 22:20:55Z - Updated version of the patch which is slightly more conformant to the style guidelines.
* [3909.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/84b/84b28e67dac3802a6dc50b12b3010df4a0a937b7/3639d1f33c7290c6d7cfc6e1a0d48a3d368abfea.patch) (10732 bytes) - added by mishok13 on 2011-03-15 00:54:30Z - Reresubmitted version of the patch

<details><summary>Searchable metadata</summary>

```
trac-id__3909 3909
type__defect defect
reporter__ivank ivank
priority__lowest lowest
milestone__ 
branch__branches_trial_decorated_methods_3909 branches/trial-decorated-methods-3909
branch_author__mishok13 mishok13
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1247207883000000 1247207883000000
changetime__1300166390000000 1300166390000000
version__None None
owner__exarkun exarkun
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|kelly commented|
|-|-|

This problem arises because for a decorated method:

```
method = decorated_test.TestA.test_something
getattr(method.im_class, method.__name__) != method
```

A workaround would be to set `newfunc.__name__ = fn.__name__` before returning from `a_decorator`. `functools.update_wrapper`, `functools.wrap` and `twisted.python.util.mergeFunctionMetadata` are alternatives depending on the version of python you are running.

```
from functools import wrap
def a_decorator(fn):
        print 'Decorating', fn
        @wrap(fn)
        def newfunc(*args, **kwargs):
                print "I am decorated."
                fn(*args, **kwargs)
        return newfunc
```


Attached is a patch to `twisted/trial/runner.py` which resolves the issue by searching the attributes of `method.im_class` for the method object.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ivank removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="kelly's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|kelly commented|
|-|-|

Original patch broke a couple of tests in twisted.trial.test.test_runner.TestRunner.  Uploaded a new version.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to kelly|
|-|-|

Nice fix! 2 comments:
 * the `object()` trick in the getattr call is probably not necessary.
 * the second hasattr call seems unnecessary (it's better to pass a second argument to the getattr call.
 * please use the `assert*` check methods instead of the `failUnless*` ones
 * some blank lines missing in the sample file.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="mishok13's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|mishok13 removed owner|
|-|-|

OK, as the original submitter didn't fix the patch, I've cleaned it up and resubmitted as 3909.patch, please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [31105]) Branching to 'trial-decorated-methods-3909'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [31106]) Apply patch

refs [#3909](https://github.com/twisted/twisted/issues/3909)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Woo, thanks!

A few stylistic things:

  1. `assertEquals` is preferred over `assertEqual`
  1. Tests should be named like `test_camelCase` rather than `test_under_scores`
  1. Likewise for other functions, `goodDecorator` rather than `good_decorator`
  1. Docstrings on the sample test methods explaining what state about them is interesting would help readability somewhat.
  1. A news fragment describing this bugfix for the NEWS file is needed too

This is all simple, so I'll take care of it and merge.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [31107]) Fix some style issues

refs [#3909](https://github.com/twisted/twisted/issues/3909)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [31110]) Merge trial-decorated-methods-3909

Author: kelly, mishok13
Reviewer: exarkun
Fixes: [#3909](https://github.com/twisted/twisted/issues/3909)

Add the ability to load test methods with names that do not match their name in
their containing class to trial's test loader.
